cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+	ccmd := &cobra.Command{
+		Use:   "hot-backup <db_name>",
+		Short: "Make a copy of the database without stopping",
+		Long: "Backup a database to file/stream without stopping the database engine. " +
+			"Backup can run from the beginning or starting from arbitrary transaction.",
+		PersistentPreRunE: cl.ConfigChain(cl.connect),
+		PersistentPostRun: cl.disconnect,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			file := io.Writer(os.Stdout)
+			params, err := prepareBackupParams(cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			udr, err := cl.immuClient.UseDatabase(cl.context, &schema.Database{DatabaseName: args[0]})
+			if err != nil {
+				return err
+			}
+			cl.context = metadata.NewOutgoingContext(cl.context, metadata.Pairs("authorization", udr.GetToken()))
+
+			if params.output != "-" {
+				f, err := cl.verifyBackupFile(params)
+				if err != nil {
+					return err
+				}
+				defer f.Close()
+				file = f
+			}
+
+			return cl.runHotBackup(file, params.startTx, params.progress)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	ccmd.Flags().StringP("output", "o", "-", "output file, \"-\" for stdout")
+	ccmd.Flags().Uint64("start-tx", 1, "Transaction ID to start from")
+	ccmd.Flags().Bool("progress", false, "show progress indicator")
+	ccmd.Flags().BoolP("append", "i", false, "append to file, if it already exists (for file output only)")
+	cmd.AddCommand(ccmd)
+}
+
+func prepareBackupParams(flags *pflag.FlagSet) (*backupParams, error) {
+	var params backupParams
+	var err error
+
+	params.output, err = flags.GetString("output")
+	if err != nil {
+		return nil, err
+	}
+	params.append, err = flags.GetBool("append")
+	if err != nil {
+		return nil, err
+	}
+	params.startTx, err = flags.GetUint64("start-tx")
+	if err != nil {
+		return nil, err
+	}
+	params.progress, err = flags.GetBool("progress")
+	if err != nil {
+		return nil, err
+	}
+
+	if params.startTx > 1 && params.append {
+		return nil, errors.New("don't use --append and --start-tx options together")
+	}
+
+	if params.output == "-" && params.append {
+		return nil, errors.New("--append option can be used only when outputting to the file")
+	}
+
+	return &params, nil
+}
+
+func (cl *commandlineHotBck) verifyBackupFile(params *backupParams) (*os.File, error) {
+	var f *os.File
+
+	_, err := os.Stat(params.output)
+	if err == nil {
+		// file exists - find last tx in file and check whether it matches the one in DB
+		if !params.append {
+			return nil, errors.New("file already exists, use --append option to append new data to file")
+		}
+		f, err = os.OpenFile(params.output, os.O_RDWR, 0)
+		if err != nil {
+			return nil, err
+		}
+		last, fileSignature, err := lastTxInFile(f)
+		if err != nil {
+			return nil, err
+		}
+		txn, err := cl.immuClient.TxByID(cl.context, last)
+		if err != nil {
+			return nil, fmt.Errorf("cannot find file's last transaction %d in database: %v", last, err)
+		}
+		if !bytes.Equal(fileSignature, txn.Header.EH) {
```

Comment:
I'd use txn.Header.Alh(), which captures the entire state of the database up to that transaction.

I'd rename fileSignature to checksum, digest or just alh (alh stands for accumulative linear hash).

Reply:
Checksum is better, will change everywhere in the code. Will change `EH` to `Alh`

Reply:
changed